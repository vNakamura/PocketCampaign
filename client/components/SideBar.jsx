SideBar = React.createClass({
  propTypes: {
    open: React.PropTypes.bool.isRequired
  },

  render() {
    const sideBarClasses = ClassNames('sideBar', {open: this.props.open});
    return (
      <div className={sideBarClasses}>
        <div className="ui vertical inverted menu">
          <a className="item">
            Item 1
          </a>
          <a className="item">
            Item 2
          </a>
          <a className="item">
            Item 3
          </a>
        </div>
      </div>
    );
  }
});
