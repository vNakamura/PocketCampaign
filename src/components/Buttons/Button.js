// @flow

import React from 'react';
import styled from 'styled-components';

import type { Theme } from '../../theme';

type Props = {
  active?: boolean,
};
type StyledProps = Props & { theme: Theme };

const StyledButton = styled.button`
  background: none;
  border: none;
  cursor: pointer;
  color: ${(props: StyledProps) => (
      props.active ? props.theme.button.activeColor : props.theme.button.color
    )};
  font-family: ${(props: StyledProps) => props.theme.fonts.display};
  font-size: 1rem;

  &:focus {
    outline: none;
  }
  &:hover,
  &:focus {
    color: ${(props: StyledProps) => props.theme.button.focusColor};
  }
`;
const handleMouseUp = (e: MouseEvent): void => {
  const target: EventTarget = e.currentTarget;
  if (target instanceof HTMLButtonElement) target.blur();
};

const Button = (props: Props) => <StyledButton {...props} onMouseUp={handleMouseUp} />;

Button.defaultProps = {
  active: false,
};

export default Button;
