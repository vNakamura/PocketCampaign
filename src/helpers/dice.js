// @flow

import React from 'react';
import droll from 'droll';
import seedrandom from 'seedrandom';
import replace from 'string-replace-to-array';

import NotationLink from '../components/Chat/NotationLink';

export const notationLimits = {
  minNumDice: 1,
  maxNumDice: 10,
  minNumSides: 2,
  maxNumSides: 100,
  minModifier: -1000,
  maxModifier: 1000,
};
export const diceNotationRegex = /\b([1-9]\d*)?d([1-9]\d*)([+-]\d+)?\b/gi;

type ResultError = {
  ok: false,
  message: string,
};
type ValidateResult = ResultError | { ok: true };
const validateRoll = (notation:string): ValidateResult => {
  const p = droll.parse(notation);
  if (!p) return { ok: false, message: 'Invalid dice notation' };
  if (p.numDice > notationLimits.maxNumDice) return { ok: false, message: 'Invalid dice notation. Too many dice!' };
  if (p.numSides < notationLimits.minNumDice) return { ok: false, message: 'Invalid dice notation. You need at least one die!' };
  if (p.numSides > notationLimits.maxNumSides) return { ok: false, message: 'Invalid dice notation. Too many sides!' };
  if (p.numSides < notationLimits.minNumSides) return { ok: false, message: 'Invalid dice notation. Too few sides!' };
  if (p.modifier > notationLimits.maxModifier) return { ok: false, message: 'Invalid dice notation. Modifier is too big!' };
  if (p.modifier < notationLimits.minModifier) return { ok: false, message: 'Invalid dice notation. Modifier is too big!' };
  return { ok: true };
};

export type ParseResult = ResultError | {
  ok: true,
  parsed: {
    numDice: number,
    numSides: number,
    modifier: number,
    minResult: number,
    maxResult: number,
    avgResult: number,
  }
};
export const parseRoll = (notation: string): ParseResult => {
  const v = validateRoll(notation);
  if (!v.ok) return v;
  const p = droll.parse(notation);
  return { ok: true, parsed: p };
};

export type RollResult = ResultError | {
  ok: true,
  rolls: number[],
  modifier: number,
  total: number,
};
export const roll = (notation: string, seed?: string): RollResult => {
  const v = validateRoll(notation);
  if (!v.ok) return v;

  if (seed !== undefined) seedrandom(seed, { global: true });
  const result = droll.roll(notation);
  return {
    ok: true,
    rolls: result.rolls,
    modifier: result.modifier,
    total: result.total,
  };
};

export const addLinksToText = (text: string) => {
  let i = 0;
  const result = replace(
    text,
    diceNotationRegex,
    (notation, numDice, numSides) => {
      i += 1;
      const v: ValidateResult = validateRoll(notation);
      if (v.ok && numSides > 1) return <NotationLink key={i} notation={notation} />;
      return notation;
    },
  );
  return result;
};
