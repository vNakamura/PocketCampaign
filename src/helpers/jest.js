import React from 'react';
import { ThemeProvider } from 'styled-components';
/* eslint-disable import/no-extraneous-dependencies */
import renderer from 'react-test-renderer';
import 'jest-styled-components';
/* eslint-enable import/no-extraneous-dependencies */
import theme from '../theme';

export const applyTheme = content => (
  <ThemeProvider theme={theme}>
    {content}
  </ThemeProvider>
);

export const renderWithTheme = content => renderer.create(
  applyTheme(content),
).toJSON();

export default {
  applyTheme,
  renderWithTheme,
};
