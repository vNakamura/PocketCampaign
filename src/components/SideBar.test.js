import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router';

import { applyTheme, renderWithTheme } from '../helpers/jest';
import { SideBar } from './SideBar';
import Button from './Buttons/Button';

const dispatch = jest.fn();
jest.useFakeTimers();

describe('components: SideBar', () => {
  const sidebar = (<MemoryRouter><SideBar
    currentUser={{
      name: 'User Name',
      key: 'someKey',
    }}
    visible
    dispatch={dispatch}
  /></MemoryRouter>);

  it('should render correctly', () => {
    const avatar = renderWithTheme(sidebar);
    expect(avatar).toMatchSnapshot();
  });

  it('should dispatch an action when close button is pressed', () => {
    const closeButton = mount(applyTheme(sidebar)).find(Button);
    closeButton.simulate('click');
    jest.runOnlyPendingTimers();
    expect(dispatch).toHaveBeenCalled();
  });

  it('should not have a close button if fixed', () => {
    const notFixed = mount(applyTheme(sidebar));
    expect(notFixed.find(Button)).toHaveLength(1);
    const fixed = mount(applyTheme(<MemoryRouter><SideBar
      currentUser={{
        name: 'User Name',
        key: 'someKey',
      }}
      visible
      fixed
      dispatch={dispatch}
    /></MemoryRouter>));
    expect(fixed.find(Button)).toHaveLength(0);
  });
});
