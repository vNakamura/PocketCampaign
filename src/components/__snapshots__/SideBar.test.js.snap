// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components: SideBar should render correctly 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-height: 3rem;
  margin-top: 0px;
}

.c3 {
  text-align: center;
  line-height: 2;
  margin: 0;
  color: #e6e5f1;
  display: inline-block;
  max-width: 250px;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  word-wrap: normal;
}

.c2 {
  -webkit-flex: 1 0 0;
  -ms-flex: 1 0 0;
  flex: 1 0 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: left;
}

.c4 {
  -webkit-flex: 1 0 0;
  -ms-flex: 1 0 0;
  flex: 1 0 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: right;
}

.c6 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  overflow-y: auto;
  -webkit-overflow-scrolling: touch;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c7 {
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  min-height: 100%;
}

.c10 {
  width: 10vmin;
  max-width: 50px;
  height: 10vmin;
  max-height: 50px;
  border-radius: 50%;
  margin: .4rem;
}

.c5 {
  background: none;
  border: none;
  cursor: pointer;
  color: #817b9e;
  font-family: Dosis;
  font-size: 1rem;
  padding: .2em .4em;
}

.c5:focus {
  outline: none;
}

.c5:hover,
.c5:focus {
  color: #e6e5f1;
}

.c8 {
  color: #e6e5f1;
  display: block;
  padding: .8rem;
  text-decoration: none;
}

.c8:hover {
  background-color: #443a53;
}

.c8 svg {
  font-size: 1.2em;
  margin-right: .5em;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  background: linear-gradient(11deg, #2a2932 25%, #1e1e24 82%);
  width: 16rem;
  min-width: 320px;
  color: #e6e5f1;
  -webkit-transform: translateX(    -0px  );
  -ms-transform: translateX(    -0px  );
  transform: translateX(    -0px  );
  -webkit-animation: iqjbZN .3s ease-out;
  animation: iqjbZN .3s ease-out;
  -webkit-animation-fill-mode: forwards;
  animation-fill-mode: forwards;
}

.c12 {
  position: absolute;
  top: 0;
  bottom: 0;
  left: 320px;
  width: 100vw;
  background-color: rgba(42,41,50,0.7);
  box-shadow: rgba(0, 0, 0, 0.3) 8px 0 16px inset;
  -webkit-animation: eMLfYp .2s ease-out;
  animation: eMLfYp .2s ease-out;
  -webkit-animation-fill-mode: forwards;
  animation-fill-mode: forwards;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c11 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

@media (max-width: 767px) {
  .c0 {
    position: fixed;
    z-index: 2;
    top: 0;
    bottom: 0;
    left: 0;
  }
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <div
      className="c2"
    />
    <h2
      className="c3"
    >
      Pocket Campaign
    </h2>
    <div
      className="c4"
    >
      <button
        className="c5"
        onClick={[Function]}
        onMouseUp={[Function]}
      >
        <svg
          fill="currentColor"
          height="1em"
          preserveAspectRatio="xMidYMid meet"
          style={
            Object {
              "color": undefined,
              "verticalAlign": "middle",
            }
          }
          viewBox="0 0 40 40"
          width="1em"
        >
          <g>
            <path
              d="m33.5 29.5q0 0.9-0.7 1.5l-3 3.1q-0.6 0.6-1.5 0.6t-1.5-0.6l-6.6-6.6-6.5 6.6q-0.7 0.6-1.6 0.6t-1.5-0.6l-3-3.1q-0.6-0.6-0.6-1.5t0.6-1.5l6.5-6.6-6.5-6.5q-0.6-0.7-0.6-1.6t0.6-1.5l3-3q0.6-0.6 1.5-0.6t1.6 0.6l6.5 6.6 6.6-6.6q0.6-0.6 1.5-0.6t1.5 0.6l3.1 3q0.6 0.7 0.6 1.5t-0.6 1.6l-6.6 6.5 6.6 6.6q0.6 0.6 0.6 1.5z"
            />
          </g>
        </svg>
      </button>
    </div>
  </div>
  <div
    className="c6"
  >
    <div
      className="c7"
    >
      <a
        className="c8"
        href="/sign-out"
        onClick={[Function]}
      >
        <svg
          fill="currentColor"
          height="1em"
          preserveAspectRatio="xMidYMid meet"
          style={
            Object {
              "color": undefined,
              "verticalAlign": "middle",
            }
          }
          viewBox="0 0 40 40"
          width="1em"
        >
          <g>
            <path
              d="m16.8 32.1q0 0.1 0 0.5t0 0.6 0 0.5-0.3 0.4-0.4 0.2h-7.2q-2.6 0-4.5-1.9t-1.9-4.5v-15.8q0-2.6 1.9-4.5t4.5-1.9h7.2q0.3 0 0.5 0.2t0.2 0.5q0 0.1 0 0.5t0 0.6 0 0.5-0.3 0.4-0.4 0.2h-7.2q-1.4 0-2.5 1t-1 2.5v15.8q0 1.4 1 2.5t2.5 1h7l0.2 0.1 0.3 0 0.2 0.1 0.1 0.2 0.1 0.3z m20.7-12.1q0 0.6-0.4 1l-12.2 12.1q-0.4 0.5-1 0.5t-1-0.5-0.4-1v-6.4h-10q-0.6 0-1-0.4t-0.4-1v-8.6q0-0.6 0.4-1t1-0.4h10v-6.4q0-0.6 0.4-1t1-0.5 1 0.5l12.2 12.1q0.4 0.4 0.4 1z"
            />
          </g>
        </svg>
        Sign Out
      </a>
      <a
        className="c8"
        href="/settings"
        onClick={[Function]}
      >
        <svg
          fill="currentColor"
          height="1em"
          preserveAspectRatio="xMidYMid meet"
          style={
            Object {
              "color": undefined,
              "verticalAlign": "middle",
            }
          }
          viewBox="0 0 40 40"
          width="1em"
        >
          <g>
            <path
              d="m25.9 20q0-2.4-1.7-4t-4.1-1.7-4 1.7-1.7 4 1.7 4 4 1.7 4.1-1.7 1.7-4z m11.4-2.4v4.9q0 0.3-0.2 0.5t-0.4 0.3l-4.2 0.6q-0.4 1.3-0.8 2.1 0.7 1.1 2.3 3.1 0.3 0.2 0.3 0.5t-0.2 0.5q-0.6 0.9-2.2 2.4t-2.1 1.6q-0.3 0-0.6-0.2l-3.1-2.4q-1 0.5-2 0.9-0.4 3-0.7 4.1-0.1 0.6-0.8 0.6h-4.9q-0.3 0-0.6-0.1t-0.2-0.5l-0.7-4.1q-1.1-0.4-2-0.9l-3.1 2.4q-0.2 0.2-0.6 0.2-0.3 0-0.5-0.2-2.8-2.6-3.7-3.8-0.2-0.2-0.2-0.5 0-0.2 0.2-0.5 0.3-0.5 1.1-1.5t1.2-1.5q-0.6-1.2-0.9-2.3l-4.1-0.6q-0.2 0-0.4-0.2t-0.2-0.6v-4.9q0-0.3 0.2-0.5t0.4-0.3l4.2-0.7q0.3-1 0.8-2-0.9-1.3-2.4-3.1-0.2-0.3-0.2-0.5 0-0.2 0.2-0.5 0.6-0.8 2.2-2.4t2.1-1.6q0.3 0 0.6 0.2l3.1 2.4q1-0.5 2-0.9 0.4-3 0.7-4.1 0.1-0.6 0.8-0.6h4.9q0.3 0 0.6 0.1t0.2 0.5l0.7 4.1q1 0.4 2 0.9l3.1-2.4q0.2-0.2 0.6-0.2 0.3 0 0.5 0.2 2.9 2.6 3.7 3.8 0.2 0.2 0.2 0.5 0 0.2-0.2 0.5-0.4 0.5-1.2 1.5t-1.2 1.5q0.6 1.2 1 2.2l4 0.7q0.3 0 0.5 0.2t0.2 0.6z"
            />
          </g>
        </svg>
        Settings
      </a>
      <a
        className="c8"
        href="/r/tutorial"
        onClick={[Function]}
      >
        <svg
          fill="currentColor"
          height="1em"
          preserveAspectRatio="xMidYMid meet"
          style={
            Object {
              "color": undefined,
              "verticalAlign": "middle",
            }
          }
          width="1em"
        />
        Introduction Tutorial
      </a>
    </div>
  </div>
  <div
    className="c9"
  >
    <img
      className="c10"
      src="https://api.adorable.io/avatars/128/someKey"
    />
    <h4
      className="c11"
    >
      User Name
    </h4>
  </div>
  <div
    className="c12"
    onClick={[Function]}
  />
</div>
`;
